File: python/kohaku/src/kohaku/registry/generation.py
"""Building the schema for L1 constrained generation (Port of TS generation.ts).

- A variant (anyOf) per component type. props are derived from the catalog's true schema.
- data $ref is pinned to an enum of the QueryHandle URIs resolved by the SemanticPort, so the LLM
  cannot forge an unknown reference (schema-level enforcement of the reference-passing principle).
- Event payloads are {key,value} pair arrays (strict mode does not allow free-form objects).

Client-local state (state / visibleWhen / emit:"state.set") and two-way binding
(data.bind / control.select) are not put in the generation vocabulary (structural containment).
"""

from __future__ import annotations

import copy
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import Any

from .catalog import ResolvedCatalog

_ID_PATTERN = "^[a-zA-Z][a-zA-Z0-9_-]{0,63}$"

GENERATION_ALWAYS_INCLUDED = ("layout.stack", "presentMarkdown")
"""Components always kept in the generation vocabulary even when candidate narrowing (include_types) is specified.

layout.stack is the root container and presentMarkdown is the terminal of the deterministic fallback;
if either disappears from the vocabulary, the L1 repair loop can no longer hold structurally.
"""


@dataclass(frozen=True)
class GeneratedDraft:
    """A draft decoded from the LLM-generated object (a wire dict before schema validation).

    The TS implementation also does not validate at this stage (the later catalog.validate + Spec parse do),
    so the Python side likewise keeps it as a dict rather than modeling it.
    """

    components: list[dict[str, Any]]
    events: list[dict[str, Any]]


@dataclass(frozen=True)
class GenerationSchema:
    jsonSchema: dict[str, Any]
    """JSON Schema already converted to the provider lowest common denominator (strict-mode compatible)"""
    decode: Callable[[object], GeneratedDraft]
    """LLM-generated object -> draft (null removal / payload pairs -> object)"""


def select_generation_types(
    catalog: ResolvedCatalog, include_types: Sequence[str] | None = None
) -> list[str]:
    """Determine the component types that actually appear in the generation vocabulary (the single source that keeps the schema and prompt vocabularies aligned).

    Decided in the order: exclude generation:"excluded" -> narrow by include_types -> guardrail union ->
    zero-count fallback.
    """
    listed = [d for d in catalog.list() if d.generation != "excluded"]
    if include_types is None:
        return [d.type for d in listed]
    # If none of the specified types exist in the catalog, fall back to all (to avoid an empty vocabulary).
    requested = [d for d in listed if d.type in include_types]
    if len(requested) == 0:
        return [d.type for d in listed]
    allowed = set(include_types) | set(GENERATION_ALWAYS_INCLUDED)
    return [d.type for d in listed if d.type in allowed]


def build_generation_schema(
    catalog: ResolvedCatalog,
    data_refs: list[str],
    include_types: Sequence[str] | None = None,
) -> GenerationSchema:
    included = set(select_generation_types(catalog, include_types))
    variants: list[dict[str, Any]] = []
    for definition in catalog.list():
        if definition.type not in included:
            continue
        props = to_generation_props_schema(definition.propsSchema.json_schema)
        properties: dict[str, Any] = {
            "id": {"type": "string", "pattern": _ID_PATTERN},
            "type": {"const": definition.type},
            "props": props,
        }
        required = ["id", "type", "props"]

        if definition.capabilities.children == "optional":
            properties["children"] = {
                "anyOf": [
                    {"type": "array", "items": {"type": "string", "pattern": _ID_PATTERN}},
                    {"type": "null"},
                ]
            }
            required.append("children")
        if definition.capabilities.data != "none" and len(data_refs) > 0:
            ref_schema = {
                "type": "object",
                "properties": {"$ref": {"type": "string", "enum": list(data_refs)}},
                "required": ["$ref"],
                "additionalProperties": False,
            }
            properties["data"] = (
                ref_schema
                if definition.capabilities.data == "required"
                else {"anyOf": [ref_schema, {"type": "null"}]}
            )
            required.append("data")

        variants.append(
            {
                "type": "object",
                "properties": properties,
                "required": required,
                "additionalProperties": False,
                "description": f"{definition.type}@{definition.version}: {definition.description}",
            }
        )

    event_schema = {
        "type": "object",
        "properties": {
            "on": {
                "type": "string",
                "description": 'componentId.eventName format (e.g. "table1.rowClick")',
            },
            "emit": {
                "type": "string",
                "enum": ["intent.patch", "intent.replace", "action.invoke"],
            },
            "payload": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {"key": {"type": "string"}, "value": {"type": "string"}},
                    "required": ["key", "value"],
                    "additionalProperties": False,
                },
            },
        },
        "required": ["on", "emit", "payload"],
        "additionalProperties": False,
    }

    json_schema: dict[str, Any] = {
        "type": "object",
        "properties": {
            "components": {"type": "array", "items": {"anyOf": variants}},
            "events": {"type": "array", "items": event_schema},
        },
        "required": ["components", "events"],
        "additionalProperties": False,
    }

    return GenerationSchema(jsonSchema=json_schema, decode=_decode)


def _decode(raw: object) -> GeneratedDraft:
    """Defensive decode: because the schema is not enforced on the prompt-JSON fallback path, an LLM's
    malformed output is thrown as a descriptive Error so the caller (l1_generate) can pick it up as a
    repair issue (rather than a TypeError turning the whole compose into INTERNAL).
    """
    obj = raw if isinstance(raw, dict) else {}
    raw_components = obj.get("components", [])
    if not isinstance(raw_components, list):
        raise ValueError("generated draft components is not an array")
    components: list[dict[str, Any]] = []
    for i, c in enumerate(raw_components):
        if not isinstance(c, dict):
            raise ValueError(f"generated draft components[{i}] is not an object")
        node = strip_nulls(c)
        assert isinstance(node, dict)
        node["props"] = node.get("props") or {}
        components.append(node)
    raw_events = obj.get("events", [])
    if not isinstance(raw_events, list):
        raise ValueError("generated draft events is not an array")
    events: list[dict[str, Any]] = []
    for i, e in enumerate(raw_events):
        if not isinstance(e, dict):
            raise ValueError(f"generated draft events[{i}] is not an object")
        on = e.get("on")
        emit = e.get("emit")
        # An out-of-enum emit is "a string but invalid", so let it through and leave it to the later
        # EventBinding validation. Here we reject only non-strings that would produce a TypeError.
        if not isinstance(on, str):
            raise ValueError(f"generated draft events[{i}].on is not a string")
        if not isinstance(emit, str):
            raise ValueError(f"generated draft events[{i}].emit is not a string")
        raw_payload = e.get("payload")
        payload_pairs = raw_payload if isinstance(raw_payload, list) else []
        payload: dict[str, Any] = {}
        for p in payload_pairs:
            pair = p if isinstance(p, dict) else {}
            payload[str(pair.get("key"))] = pair.get("value")
        events.append({"on": on, "emit": emit, "payload": payload})
    return GeneratedDraft(components=components, events=events)


def to_generation_props_schema(schema: dict[str, Any]) -> dict[str, Any]:
    """Deterministic conversion into the schema presented to the LLM (aligned to the provider lowest common denominator for structured output):

    - object: make every property required and turn originally-optional ones into anyOf [orig, null]
      (for OpenAI strict mode; a two-stage approach of removing null after generation and re-validating against the true schema)
    - force additionalProperties: false
    - remove default / $schema (for Gemini responseSchema compatibility)
    """
    result = _walk(copy.deepcopy(schema))
    assert isinstance(result, dict)
    return result


def _walk(node: object) -> object:
    if isinstance(node, list):
        return [_walk(v) for v in node]
    if not isinstance(node, dict):
        return node
    n = dict(node)
    n.pop("default", None)
    n.pop("$schema", None)

    if n.get("type") == "object" and isinstance(n.get("properties"), dict):
        props = n["properties"]
        required = set(n.get("required") or [])
        new_props: dict[str, Any] = {}
        for key, value in props.items():
            transformed = _walk(value)
            new_props[key] = (
                transformed if key in required else {"anyOf": [transformed, {"type": "null"}]}
            )
        n["properties"] = new_props
        n["required"] = list(new_props.keys())
        n["additionalProperties"] = False
        return n

    for key in ("items", "anyOf", "oneOf", "allOf", "prefixItems"):
        if n.get(key) is not None:
            n[key] = _walk(n[key])
    return n


def strip_nulls(value: object) -> object:
    """Remove null-valued properties from the generation result (the "null = omitted" convention). A precursor to re-validation."""
    if isinstance(value, list):
        return [strip_nulls(v) for v in value]
    if isinstance(value, dict):
        return {k: strip_nulls(v) for k, v in value.items() if v is not None}
    return value

File: packages/registry/src/generation.ts
import type { ComponentNode, EventBinding, JsonObject, JsonValue } from "@kohaku-ui/spec-core";
import type { ResolvedCatalog } from "./catalog.js";
import { cachedGenerationPropsSchema, stripNulls } from "./json-schema.js";

export interface GeneratedDraft {
  components: ComponentNode[];
  events: EventBinding[];
}

export interface GenerationSchema {
  /** JSON Schema already converted to the provider lowest common denominator (strict-mode compatible) */
  jsonSchema: Record<string, unknown>;
  /** LLM-generated object → draft (null removal / payload pairs → object) */
  decode(raw: unknown): GeneratedDraft;
}

const ID_PATTERN = "^[a-zA-Z][a-zA-Z0-9_-]{0,63}$";

/**
 * Parts that always remain in the generation vocabulary even when candidate narrowing (includeTypes)
 * is specified. layout.stack is the root container and presentMarkdown is the terminal of the
 * deterministic fallback; if either disappears from the vocabulary, the L1 repair loop no longer holds
 * structurally (the root cannot be built / a fallback Spec cannot be produced). Always unioned in as a
 * guardrail.
 */
const GENERATION_ALWAYS_INCLUDED = ["layout.stack", "presentMarkdown"] as const;

/** Optional settings for buildGenerationSchema. */
export interface BuildGenerationOptions {
  /**
   * Allowlist of component types to include in the generation vocabulary (candidate narrowing).
   * - undefined: all of them (the default, excluding generation:"excluded").
   * - when specified: only these types + the guardrails (layout.stack / presentMarkdown), always unioned.
   * - if none of the specified types exists in the catalog, fall back to all of them (prevents an empty
   *   vocabulary that would make repair impossible).
   */
  includeTypes?: readonly string[];
}

/**
 * Determines the component types that actually appear in the generation vocabulary (the single source
 * that keeps buildGenerationSchema and the prompt in agreement on the vocabulary).
 * Finalized in this order: exclude generation:"excluded" → narrow by includeTypes → union guardrails →
 * fall back on zero matches.
 */
export function selectGenerationTypes(catalog: ResolvedCatalog, includeTypes?: readonly string[]): string[] {
  const listed = catalog.list().filter((def) => def.generation !== "excluded");
  if (includeTypes == null) return listed.map((def) => def.type);
  // If none of the specified types exists in the catalog, fall back to all of them (avoids an empty vocabulary).
  const requested = listed.filter((def) => includeTypes.includes(def.type));
  if (requested.length === 0) return listed.map((def) => def.type);
  const allowed = new Set<string>([...includeTypes, ...GENERATION_ALWAYS_INCLUDED]);
  return listed.filter((def) => allowed.has(def.type)).map((def) => def.type);
}

/**
 * Dynamically builds the schema for L1 constrained generation.
 * - A variant per component type (anyOf). props are derived from the catalog's true schema.
 * - The data $ref is pinned to an enum of QueryHandle URIs resolved by the SemanticPort so the LLM
 *   cannot forge unknown references (schema-level enforcement of the reference-passing principle).
 * - The event payload is a {key,value} pair array (strict mode disallows free-form objects).
 *
 * Client-local state (kohaku >= 0.2 state / visibleWhen / emit:"state.set") and two-way binding
 * (data.bind / control.select) are **not put into the generation vocabulary**. Because a node variant
 * is additionalProperties:false, neither visibleWhen nor data.bind can be output structurally (data is
 * `{$ref: enum}` + additionalProperties:false, which seals off the bind keys), and the event emit enum
 * does not include state.set either. control.select is generation:"excluded" and drops out of the
 * variants. Likewise, the overlay parts (overlay.dialog / overlay.toast) are state-linked parts that
 * declare their open/close via state + visibleWhen, so they are generation:"excluded" (dropped from the
 * variants). These are intended for L0 fixed Specs, hand-written Specs, and promotion templates;
 * opening them to L1 generation is left as a separate decision (a structural containment that does not
 * let the LLM decide the bind value-range allowlist).
 */
export function buildGenerationSchema(
  catalog: ResolvedCatalog,
  dataRefs: string[],
  options?: BuildGenerationOptions,
): GenerationSchema {
  // In addition to excluding generation:"excluded" (runtime-only parts such as ui.loading), narrow the
  // candidates when includeTypes is specified. selectGenerationTypes is the single source of the types
  // that actually remain (keeping prompt and vocabulary in agreement). Combined with
  // additionalProperties:false, the LLM cannot output anything other than these structurally.
  const included = new Set(selectGenerationTypes(catalog, options?.includeTypes));
  const variants = catalog
    .list()
    .filter((def) => included.has(def.type))
    .map((def) => {
      const props = cachedGenerationPropsSchema(def);
      const properties: Record<string, unknown> = {
        id: { type: "string", pattern: ID_PATTERN },
        type: { const: def.type },
        props,
      };
      const required = ["id", "type", "props"];

      if (def.capabilities.children === "optional") {
        properties["children"] = {
          anyOf: [{ type: "array", items: { type: "string", pattern: ID_PATTERN } }, { type: "null" }],
        };
        required.push("children");
      }
      if (def.capabilities.data !== "none" && dataRefs.length > 0) {
        const refSchema = {
          type: "object",
          properties: { $ref: { type: "string", enum: dataRefs } },
          required: ["$ref"],
          additionalProperties: false,
        };
        properties["data"] =
          def.capabilities.data === "required" ? refSchema : { anyOf: [refSchema, { type: "null" }] };
        required.push("data");
      }

      return {
        type: "object",
        properties,
        required,
        additionalProperties: false,
        description: `${def.type}@${def.version}: ${def.description}`,
      };
    });

  const eventSchema = {
    type: "object",
    properties: {
      on: { type: "string", description: 'componentId.eventName format (e.g. "table1.rowClick")' },
      emit: { type: "string", enum: ["intent.patch", "intent.replace", "action.invoke"] },
      payload: {
        type: "array",
        items: {
          type: "object",
          properties: { key: { type: "string" }, value: { type: "string" } },
          required: ["key", "value"],
          additionalProperties: false,
        },
      },
    },
    required: ["on", "emit", "payload"],
    additionalProperties: false,
  };

  const jsonSchema = {
    type: "object",
    properties: {
      components: { type: "array", items: { anyOf: variants } },
      events: { type: "array", items: eventSchema },
    },
    required: ["components", "events"],
    additionalProperties: false,
  };

  return {
    jsonSchema,
    decode(raw: unknown): GeneratedDraft {
      // Defensive validation: on the prompt-JSON fallback path (the default
      // KOHAKU_LLM_STRUCTURED_MODE=auto), jsonSchema is not enforced, so malformed LLM output
      // (components/events not an array, an element not an object, on/emit not a string) reaches decode
      // unvalidated. Throw a descriptive Error here so the caller (l1-generate) can pick it up as a
      // repair issue (a bare .map / property access would raise a TypeError, which would turn the whole
      // compose into INTERNAL / a host 500).
      const obj = (raw ?? {}) as { components?: unknown; events?: unknown };
      const rawComponents = obj.components ?? [];
      if (!Array.isArray(rawComponents)) {
        throw new Error("generated draft components is not an array");
      }
      const components = rawComponents.map((c, i): ComponentNode => {
        if (c == null || typeof c !== "object" || Array.isArray(c)) {
          throw new Error(`generated draft components[${i}] is not an object`);
        }
        const node = stripNulls(c) as ComponentNode;
        return { ...node, props: (node.props ?? {}) as JsonObject };
      });
      const rawEvents = obj.events ?? [];
      if (!Array.isArray(rawEvents)) {
        throw new Error("generated draft events is not an array");
      }
      const events: EventBinding[] = rawEvents.map((e, i): EventBinding => {
        if (e == null || typeof e !== "object" || Array.isArray(e)) {
          throw new Error(`generated draft events[${i}] is not an object`);
        }
        const ev = e as { on?: unknown; emit?: unknown; payload?: unknown };
        // An out-of-enum emit (e.g. "intent.foo") is "a string but invalid", so decode lets it through
        // and defers to the EventBinding Zod validation in collectIssues. Here we reject only
        // non-strings that would raise a TypeError.
        if (typeof ev.on !== "string") {
          throw new Error(`generated draft events[${i}].on is not a string`);
        }
        if (typeof ev.emit !== "string") {
          throw new Error(`generated draft events[${i}].emit is not a string`);
        }
        const payload = Array.isArray(ev.payload) ? ev.payload : [];
        return {
          on: ev.on,
          emit: ev.emit as EventBinding["emit"],
          payload: Object.fromEntries(
            payload.map((p) => {
              const pair = (p ?? {}) as { key?: unknown; value?: unknown };
              return [String(pair.key), pair.value as JsonValue];
            }),
          ),
        };
      });
      return { components, events };
    },
  };
}
